src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java

```diff
@@ -93,4 +99,139 @@ public static String toQueryString(Map<String, String> params) {
+    }
+
+    /**
+     * Check whether a new leg of an itinerary matches the previous itinerary leg for the purposes of trip monitoring.
+     */
+    private static boolean legsMatch(Leg previousItineraryLeg, Leg newItineraryLeg) {
+        // for now don't analyze non-transit legs
+        if (!previousItineraryLeg.transitLeg) return true;
+
+        // make sure the same from/to stop are being used
+        if (
+            !stopsMatch(previousItineraryLeg.from, newItineraryLeg.from) ||
+                !stopsMatch(previousItineraryLeg.to, newItineraryLeg.to)
+        ) {
+            return false;
+        }
+
+        // make sure the transit service is the same as perceived by the customer
+        if (
+            !equalsOrPreviousWasNull(previousItineraryLeg.mode, newItineraryLeg.mode) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(previousItineraryLeg.agencyName, newItineraryLeg.agencyName) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(
+                    previousItineraryLeg.routeLongName,
+                    newItineraryLeg.routeLongName
+                ) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(
+                    previousItineraryLeg.routeShortName,
+                    newItineraryLeg.routeShortName
+                ) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(
+                    previousItineraryLeg.headsign,
+                    newItineraryLeg.headsign
+                ) ||
+                !equalsOrPreviousWasNull(
+                    previousItineraryLeg.interlineWithPreviousLeg,
+                    newItineraryLeg.interlineWithPreviousLeg
+                )
+        ) {
+            return false;
+        }
+
+        // make sure the transit trips are scheduled for the same time of the day
+        if (
+            !timeOfDayMatches(
+                previousItineraryLeg.getScheduledStartTime(),
+                newItineraryLeg.getScheduledStartTime()
+            ) || !timeOfDayMatches(
+                previousItineraryLeg.getScheduledEndTime(),
+                newItineraryLeg.getScheduledEndTime()
+            )
+        ) {
+            return false;
+        }
+
+        // if this point is reached, the legs are assumed to match
+        return true;
+    }
+
+    /**
+     * Checks whether two stops (OTP Places) match for the purposes of matching itineraries
+     */
+    private static boolean stopsMatch(Place stopA, Place stopB) {
+        // stop names must match
+        if (!stopA.name.equalsIgnoreCase(stopB.name)) return false;
+
+        // stop code must match
+        if (!equalsIgnoreCaseOrPreviousWasEmpty(stopA.stopCode, stopB.stopCode)) return false;
+
+        // stop positions must be no further than 5 meters apart
+        if (
+            DistanceUtils.radians2Dist(
+                DistanceUtils.distHaversineRAD(stopA.lat, stopA.lon, stopB.lat, stopB.lon),
+                DistanceUtils.EARTH_MEAN_RADIUS_KM
+            ) * 1000 > 5
```

Comment:
I'd prefer some variable names in this method's conditionals. E.g.:

```java
double stopDistanceMeters = DistanceUtils.radians2Dist(
                DistanceUtils.distHaversineRAD(stopA.lat, stopA.lon, stopB.lat, stopB.lon),
                DistanceUtils.EARTH_MEAN_RADIUS_KM
            ) * 1000;
if (stopDistanceMeters > 5)
....

```

Reply:
Fixed in ffc3b49b3ad576d936b04758d572efdf3f217ab2